File: vagrant/hack/lib/cluster-api.rb
require 'yaml'

require_relative 'kubeadm-options.rb'
require_relative 'kubernetes-utils.rb'
require_relative 'vagrant-utils.rb'
require_relative 'ansible-utils.rb'
require_relative 'utils.rb'

require_relative 'test-guide.rb'

class VagrantMachine
    attr_accessor :name
    attr_accessor :hostname
    attr_accessor :box
    attr_accessor :ip
    attr_accessor :cpus
    attr_accessor :memory
    attr_accessor :roles
end

class VagrantMachineSet
    attr_accessor :name
    attr_accessor :replicas
    attr_accessor :box
    attr_accessor :cpus
    attr_accessor :memory
    attr_accessor :roles
end

class VagrantCluster
    
    attr_reader :name
    attr_reader :version
    attr_reader :controlplane
    attr_reader :controlplaneAvailability
    attr_reader :etcdMode
    attr_reader :certificateAuthority
    attr_reader :pkiLocation
    attr_reader :dnsAddon
    attr_reader :dnsDomain
    attr_reader :networkAddon
    attr_reader :serviceSubnet
    attr_reader :podSubnet
    attr_reader :kubeletConfig
    
    attr_reader :machines
    attr_reader :masters
    attr_reader :nodes
    attr_reader :etcds

    attr_reader :ansible_installEtcd
    attr_reader :ansible_setupExternalCA
    attr_reader :ansible_setupControlplaneVip
    attr_reader :ansible_kubeadmInit
    attr_reader :ansible_applyNetwork
    attr_reader :ansible_kubeadmJoinMaster
    attr_reader :ansible_kubeadmJoin

    attr_reader :ansible_groups
    attr_reader :ansible_extra_vars
    attr_reader :ansible_tags

    def self.fromClusterAPI (folder)
        c = VagrantCluster.new
        c.parseFolder folder
        c.compose
        return c
    end

    #$ ruby -r "./hack/lib/cluster-api.rb" -e "VagrantCluster.Test" up
    def self.Test
        c = fromClusterAPI('/Users/fabrizio/Documents/Workspace/Vagrant/kubernetes/dev/spec/*.yaml')
            .injectKubeadmFromBazelBuild(prefix: nil)
            .ansibleActions(
                installEtcd: true,              # only if machine with Etcd role are defined (external etcd)
                setupExternalCA: true,          # only if external certificate is requested
                setupControlplaneVip: true,     # only if there are 2 or more machines with Master role (multi master scenario)
                kubeadmInit: false, 
                applyNetwork: false,
                kubeadmJoinMaster: false, 
                kubeadmJoin: false
            )

        TestGuide.print(c)
    end

    def initialize
        @machineSets            = []
        @machines               = []
        @silent = VagrantUtils.silent
    end

    def parseFolder(folder)
        Utils.putsTitle "Reading cluster API spec from #{folder}" unless @silent
        Dir[folder].each do |file_name|
            next if File.directory? file_name 

            puts  "    - #{file_name}" unless @silent
            data = YAML.load_file(file_name)
  
            case data["kind"]
            when "Cluster"
                parseClusterData data
            when "MachineSet"
                parseMachineSetData data
            end
        end
    end

    def parseClusterData(data)
        @name                       = get data, "metadata", "name", validator: KuberentesUtils::DNS1123Label 
        @extra_vars                 = get data, "spec", "providerConfig", "value"

        @version                    = get data, "spec", "providerConfig", "value", "kubernetes", "version" #TODO: validate versione
        @controlplane               = get data, "spec", "providerConfig", "value", "kubernetes", "controlplane", default: KubeadmOptions::Controlplane::StaticPods, validator: KubeadmOptions::Controlplane
        # controlplaneAvailability will be implicitily derived from machineSets during compose
        # etcdMode will be implicitily derived from machineSets during compose
        @certificateAuthority       = get data, "spec", "providerConfig", "value", "kubernetes", "certificateAuthority", default: KubeadmOptions::CertificateAuthority::Local, validator: KubeadmOptions::CertificateAuthority
        @pkiLocation                = get data, "spec", "providerConfig", "value", "kubernetes", "pkiLocation", default: KubeadmOptions::PKILocation::Filesystem, validator: KubeadmOptions::PKILocation
        @dnsAddon                   = get data, "spec", "providerConfig", "value", "kubernetes", "dnsAddon", default: KubeadmOptions::DNSAddon::KubeDNS, validator: KubeadmOptions::DNSAddon
        @dnsDomain                  = get data, "spec", "clusterNetwork", "serviceDomain", default: KubeadmOptions::DNSDomain , validator: KuberentesUtils::DNS1123Subdomain
        @networkAddon               = get data, "spec", "providerConfig", "value", "kubernetes", "networkAddon", default: KubeadmOptions::NetworkAddon::Weavenet, validator: KubeadmOptions::NetworkAddon
        serviceSubnet, podSubnet    = KubeadmOptions::NetworkAddon.defaultSubnetsFor @networkAddon
        @serviceSubnet              = get data, "spec", "clusterNetwork", "services", "cidrblocks", default: serviceSubnet, validator: Utils::CIDRList
        @podSubnet                  = get data, "spec", "clusterNetwork", "pods", "cidrblocks", default: podSubnet, validator: Utils::CIDRList
        @kubeletConfig              = get data, "spec", "providerConfig", "value", "kubernetes", "kubeletConfig", default: KubeadmOptions::KubeletConfig::SystemdDropIn, validator: KubeadmOptions::KubeletConfig
        
        if @pkiLocation == KubeadmOptions::PKILocation::Secrets and @controlplane != KubeadmOptions::Controlplane::SelfHosted then raise "Invalid cluster definition. PKILocation can be secrets only if controlplane is self hosted" end
    end

    def parseMachineSetData(data)
        s = VagrantMachineSet.new 
        s.name        = get data, "metadata", "name", validator: KuberentesUtils::DNS1123Label
        s.replicas    = get data, "spec", "replicas", validator: Utils::Integer
        s.box         = get data, "spec", "template", "spec", "providerConfig", "value", "box"
        s.cpus        = get data, "spec", "template", "spec", "providerConfig", "value", "cpus", validator: Utils::Integer
        s.memory      = get data, "spec", "template", "spec", "providerConfig", "value", "memory", validator: Utils::Integer
        s.roles       = get data, "spec", "template", "spec", "roles", validator: Roles
        @machineSets << s
    end

    def compose
        @masters = []
        @etcds = []
        @nodes = []

        @machineSets.each do |s|
            for i in 1..s.replicas
                m = VagrantMachine.new
                m.name              = if s.replicas == 1 then "#{@name}-#{s.name}" else "#{@name}-#{s.name}#{i}" end  
                m.hostname          = "#{@name}-#{s.name}.local"
                m.box               = s.box
                m.ip                = "10.10.10.1#{i}"
                m.cpus              = s.cpus
                m.memory            = s.memory
                m.roles             = s.roles
                machines << m

                @masters << m if m.roles.include?(Roles::Master)
                @nodes   << m if m.roles.include?(Roles::Node)
                @etcds   << m if m.roles.include?(Roles::Etcd)
            end
        end

        if @masters.length == 0 then raise "Invalid cluster definition. At least one Master machine is required" end
        
        @etcdMode = KubeadmOptions::EtcdMode::Local
        if @etcds.length > 0 then @etcdMode = KubeadmOptions::EtcdMode::External end

        @controlplaneAvailability = KubeadmOptions::ControlplaneAvailability::SingleMaster 
        if @masters.length > 1 then 
            @controlplaneAvailability = KubeadmOptions::ControlplaneAvailability::MultiMaster 
            if @etcdMode != KubeadmOptions::EtcdMode::External then raise "Invalid cluster definition. Multi masters requires external etcd." end
            if @pkiLocation == KubeadmOptions::PKILocation::Secrets then raise "Invalid cluster definition. Multi masters does not support certificates in secrets yet." end
        end
        
        Utils.putsTitle "Machines defined in cluster API spec" unless @silent
        @machines.each do |m|
            puts "    - #{m.name}, roles: #{m.roles.join(",")} ]" unless @silent
        end
    end

    def get(data, *keys, default: nil, validator: nil)
        d = data
        keys.each do |k| 
            if d.has_key?(k)
                d = d[k]
            else
                if default!=nil 
                    return default
                else
                    raise "Error parsing #{keys} (#{k} does't exists)"
                end
            end
        end
        if validator!=nil 
            validator.validate d
        end
        return d
    end

    def injectKubeadmFromBazelBuild(prefix: nil)
        @ansible_extra_vars['kubeadm_binary'] = VagrantUtils::InjectKubeadmFromBuildOutput(build_output_path = "bazel-bin/cmd/kubeadm/linux_amd64_pure_stripped", prefix = prefix)
        return self
    end

    def injectKubeadmFromDockerBuild(prefix: nil)
        @ansible_extra_vars['kubeadm_binary'] = VagrantUtils::InjectKubeadmFromBuildOutput(build_output_path = "_output/dockerized/bin/linux/amd64/", prefix = prefix)
        return self
    end

    def injectKubeadmFromLocalBuild(prefix: nil)
        @ansible_extra_vars['kubeadm_binary'] = VagrantUtils::InjectKubeadmFromBuildOutput(build_output_path = "_output/local/bin/linux/amd64/", prefix = prefix)
        return self
    end

    def ansibleActions(installEtcd: true, setupExternalCA: true, setupControlplaneVip: true, kubeadmInit: false, applyNetwork: false, kubeadmJoinMaster: false, kubeadmJoin: false)    
        @ansible_installEtcd            = installEtcd
        @ansible_setupExternalCA        = setupExternalCA
        @ansible_setupControlplaneVip   = setupControlplaneVip
        @ansible_kubeadmInit            = kubeadmInit
        @ansible_applyNetwork           = applyNetwork
        @ansible_kubeadmJoinMaster      = kubeadmJoinMaster
        @ansible_kubeadmJoin            = kubeadmJoin

        @ansible_groups = AnsibleUtils::Inventory.for @machines, Roles::All
        @ansible_extra_vars = @extra_vars
        @ansible_extra_vars['kubernetes']['dnsDomain']     = @dnsDomain            unless @dnsDomain == KubeadmOptions::DNSDomain
        @ansible_extra_vars['kubernetes']['serviceSubnet'] = @serviceSubnet.first  unless @serviceSubnet.first == nil
        @ansible_extra_vars['kubernetes']['podSubnet']     = @podSubnet.first      unless @podSubnet.first == nil
        @ansible_tags   = AnsibleUtils::Tags.for installEtcd, setupExternalCA, setupControlplaneVip, kubeadmInit, applyNetwork, kubeadmJoinMaster, kubeadmJoin
        return self
    end
end

class Roles
    Master  = 'Master'    
    Node    = 'Node'
    Etcd    = 'Etcd'
    
    All    = [Master, Node, Etcd]
    
    def self.validate(data)
        if data.kind_of?(Array)
            data.each do |s|
                raise "Invalid role '#{s}'. valid roles are [#{All.join(', ')}]." unless All.include?(s)
            end 
        else
            raise "invalid roles list '#{data}'" 
        end
    end
end 

File: vagrant/hack/lib/cluster_api.py
# -*- coding: utf-8 -*-

# Copyright 2018 The Kubernetes Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
import sys
import glob
import yaml

import utils
import kubeadm_utils
import vagrant_utils
import kubernetes_utils

ROLE_MASTER  = 'Master'      
ROLE_NODE    = 'Node' 
ROLE_ETCD    = 'Etcd'

def validate_roles(args):
    valid = [ROLE_MASTER, ROLE_NODE, ROLE_ETCD]
    for arg in args:
        if arg not in valid:
            raise ValueError("invalid machine set role '%s'. Valid roles are %s" % (arg, ', '.join(valid))) 

class VagrantCluster:
    """ represent a VagrantCluster, as defined in the cluster API  """

    def __init__(self):
 
        self.name = ""
        self.version = ""
        self.controlplane = ""
        self.certificateAuthority = ""
        self.pkiLocation = ""
        self.dnsAddon = ""
        self.dnsDomain = ""                 
        self.networkAddon = ""              
        self.serviceSubnet = ""             
        self.podSubnet = ""
        self.kubeletConfig = ""
        self.extra_vars = {}                # the whole providerConfig.value is used as extra_vars for ansible 
        self.highavailability = False       # nb derived from machine sets (Master machines > 1)
        self.externalEtcd = False           # nb derived from machine sets (Etcd machines > 0)

class VagrantMachineSet:
    """ represent a VagrantMachineSet, as defined in the cluster API  """

    def __init__(self):
        self.name       = ""
        self.replicas   = 0
        self.box        = ""
        self.cpus       = 0
        self.memory     = 0
        self.roles      = []

class VagrantMachine:
    """ represent a VagrantMachine creating from a VagrantMachineSet specification """

    def __init__(self):
        self.name       = ""
        self.hostname   = ""
        self.box        = ""
        self.ip         = ""
        self.cpus       = 0
        self.memory     = 0
        self.roles      = []

def parse(folder):
    """ parses the cluster api definition in a folder """
    if not os.path.isabs(folder):
        folder = os.path.join(vagrant_utils.root_folder, folder)

    cluster = None
    machineSets = []

    for f in glob.glob(folder):
        if os.path.isfile(f):
            try:           
                data = yaml.load(open(f))

                if data["kind"] == "Cluster":
                    cluster = get_cluster(data)
                elif data["kind"] == "MachineSet":
                    machineSets.append(get_machineSet(data))
            except Exception as e:
                raise RuntimeError('Error parsing cluster api specification in `' + repr(f) + '`: ' + repr(e))

    # fails fast if incomplete configurations are detected
    if cluster == None: 
        raise ValueError("Invalid cluster api specification in %s. Cluster object not defined" % (spec))
    
    if len(machineSets) == 0: 
        raise ValueError("Invalid cluster api specification in %s. MachineSets objects not defined" % (spec))

    return cluster, machineSets

def get_cluster(data):
    """ get the cluster definition from a cluster api object """

    c = VagrantCluster()

    c.name                       = getx(data, "metadata.name", validator=kubernetes_utils.validate_rfc123_label)
    c.version                    = getx(data, "spec.providerConfig.value.kubernetes.version") #TODO: validate version
    c.controlplane               = getx(data, "spec.providerConfig.value.kubernetes.controlplane", default=kubeadm_utils.CONTROLPLANE_TYPE_STATICPODS, validator=kubeadm_utils.validate_controlplane_type)
    c.certificateAuthority       = getx(data, "spec.providerConfig.value.kubernetes.certificateAuthority", default=kubeadm_utils.CERTIFICATEAUTHORITY_TYPE_LOCAL, validator=kubeadm_utils.validate_certificateAuthority_type)
    c.pkiLocation                = getx(data, "spec.providerConfig.value.kubernetes.pkiLocation", default=kubeadm_utils.CERTIFICATEAUTHORITY_LOCATION_FILESYSTEM, validator=kubeadm_utils.validate_certificateAuthority_location)
    c.dnsAddon                   = getx(data, "spec.providerConfig.value.kubernetes.dnsAddon", default=kubeadm_utils.DNS_ADDON_KUBE_DNS, validator=kubeadm_utils.validate_dns_addon)
    c.dnsDomain                  = getx(data, "spec.clusterNetwork.serviceDomain", default=kubeadm_utils.DEFAULT_DNS_DOMAIN , validator=kubernetes_utils.validate_rfc123_subdomain)
    c.networkAddon               = getx(data, "spec.providerConfig.value.kubernetes.cni.plugin", default=kubeadm_utils.NETWORK_ADDON_WEAVENET, validator=kubeadm_utils.validate_network_addon)
    (serviceSubnet, podSubnet) = kubeadm_utils.network_addon_default_subnet(c.networkAddon)
    c.serviceSubnet              = getx(data, "spec.clusterNetwork.services.cidrblocks", default=serviceSubnet, validator=utils.validate_cidr)
    c.podSubnet                  = getx(data, "spec.clusterNetwork.pods.cidrblocks", default=podSubnet, validator=utils.validate_cidrs)
    c.kubeletConfig              = getx(data, "spec.providerConfig.value.kubernetes.kubeletConfig", default=kubeadm_utils.KUBELET_CONFIG_TYPE_SYSTEMDDROPIN, validator=kubeadm_utils.validate_kubelet_config_type)

    # uses the providerConfig as an input for ansible
    c.extra_vars                 = getx(data, "spec.providerConfig.value")
    
    # values defined in cluster API attributes, are replicate under extra_vars for simplifying ansible integration
    if c.dnsDomain != kubeadm_utils.DEFAULT_DNS_DOMAIN:
        c.extra_vars['kubernetes']['dnsDomain'] = c.dnsDomain

    if c.serviceSubnet[0] != None:
        c.extra_vars['kubernetes']['serviceSubnet'] = c.serviceSubnet[0]

    if c.podSubnet[0] != None:
        c.extra_vars['kubernetes']['podSubnet']     = c.podSubnet[0]

    # fails fast if not consistent configurations are detected
    if c.pkiLocation == kubeadm_utils.CERTIFICATEAUTHORITY_LOCATION_SECRETS and c.controlplane != kubeadm_utils.CONTROLPLANE_TYPE_SELFHOSTING: 
        raise ValueError("Invalid cluster definition. PKILocation can be secrets only if controlplane is self hosted")

    return c

def get_machineSet(data):
    """ get the machine set definition from a cluster api object """

    s = VagrantMachineSet()

    s.name        = getx(data, "metadata.name", validator=kubernetes_utils.validate_rfc123_label)
    s.replicas    = getx(data, "spec.replicas", default=1, validator=utils.validate_integer)
    s.box         = getx(data, "spec.template.spec.providerConfig.value.box", default="ubuntu/xenial64") # TODO:validator
    s.cpus        = getx(data, "spec.template.spec.providerConfig.value.cpus", default=2, validator=utils.validate_integer)
    s.memory      = getx(data, "spec.template.spec.providerConfig.value.memory", default=2048, validator=utils.validate_integer)
    s.roles       = getx(data, "spec.template.spec.roles", validator=validate_roles)
   
    return s

def getx(data, keys, default=None, validator=None):
    """ extended get of an attribute of the cluster api with, recoursion (deep get), defaulting & validation """

    for key in keys.split('.'):
        try:
            data = data[key]
        except KeyError:
            if default != None:
                return default
            else:
                raise KeyError("invalid cluster api definition. Key '%s' does not exist" % (keys)) 
    
    if validator != None:
        validator(data)

    return data

def get_machines(cluster, machineSets):
    """ transform a cluster / a list of machineSets into a set of machines to be created.
        the list of machine is also stored into `tmp/machines.yaml` for vagrant execution """

    machines = []
    j = 1
    for s in machineSets:
        for i in range(1, s.replicas + 1):
            m = VagrantMachine()
            m.name              = "%s-%s" % (cluster.name, s.name) if s.replicas == 1 else "%s-%s%s" % (cluster.name, s.name, j)
            m.hostname          = "%s-%s.local" % (cluster.name, m.name)
            m.box               = s.box
            m.ip                = "10.10.10.1%s" % (j)
            m.cpus              = s.cpus
            m.memory            = s.memory
            m.roles             = s.roles
            machines.append(m)
            j += 1

    masters = [m for m in machines if ROLE_MASTER in m.roles]
    if len(masters) == 0:
        raise ValueError("Invalid cluster definition. At least one Master machine is required")
    elif len(masters) > 1:
        cluster.highavailability = True
        if sum([1 for m in machines if ROLE_ETCD in m.roles])==0:
            raise ValueError("Invalid cluster definition. Multi masters requires external etcd.")
        if cluster.pkiLocation == kubeadm_utils.CERTIFICATEAUTHORITY_LOCATION_SECRETS:
            raise ValueError("Invalid cluster definition. Multi masters does not support certificates in secrets yet.")

    etcds = [m for m in machines if ROLE_ETCD in m.roles]
    if len(etcds) > 0:
        cluster.externalEtcd = True

    if not os.path.exists(vagrant_utils.tmp_folder):
        os.makedirs(vagrant_utils.tmp_folder)

    machines_target = os.path.join(vagrant_utils.tmp_folder, 'machines.yml')
    
    with open(machines_target, 'w') as outfile:
        yaml.dump(machines, outfile, default_flow_style=False)

    return machines

def fallbackSettings(cluster, machineSets):
    """ Fixes cluster and machineSets options in case of ansible not available,
        ensuring compatibility with the fallback_boostrap script defined in the Vagrant file """ 

    for s in machineSets:
        s.box = "bento/ubuntu-17.10" # base box tested with the boostrap script
